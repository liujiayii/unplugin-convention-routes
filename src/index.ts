import type { UnpluginFactory } from 'unplugin'
import type { Options, UserOptions } from './core/types'

import path from 'node:path'

import process from 'node:process'
import { createUnplugin } from 'unplugin'
import { PageContext } from './core/context'

export const unpluginFactory: UnpluginFactory<Options | undefined> = (userOptions: UserOptions = {}) => {
  const ctx: PageContext = new PageContext(userOptions)
  //  ctx.setLogger(config.logger)
  return ({
    name: 'unplugin-pages',
    buildEnd() {
      ctx.watcher?.close()
    },
    resolveId(id) {
      //  console.log('resolveId', id)
      if (['~unplugin-pages/react', '~unplugin-pages/vue', '~unplugin-pages/solid'].includes(id)) {
        return path.resolve(process.cwd(), 'node_modules', '~unplugin-pages', `${id.split('/')[1]}.js`)
      }
      return null
    },
  })
}

export const unplugin = /* #__PURE__ */ createUnplugin(unpluginFactory)

export default unplugin
