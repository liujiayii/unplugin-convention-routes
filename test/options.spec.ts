import { resolveOptions } from '../src/options'

describe('options resolve', () => {
  it('vue', () => {
    const options = resolveOptions({
      dirs: 'examples/vue/src/pages',
    })
    expect(options).toMatchSnapshot({
      root: expect.any(String),
    })
  })

  it('vue - custom module id', () => {
    const options = resolveOptions({
      dirs: 'examples/vue/src/pages',
      moduleId: '~vue-pages',
    })
    expect(options).toMatchSnapshot({
      root: expect.any(String),
    })
  })

  it('react', () => {
    const options = resolveOptions({
      dirs: 'examples/react/src/pages',
      resolver: 'react',
    })
    expect(options).toMatchSnapshot({
      root: expect.any(String),
    })
  })

  it('solid', () => {
    const options = resolveOptions({
      dirs: 'examples/solid/src/pages',
      resolver: 'solid',
    })
    expect(options).toMatchSnapshot({
      root: expect.any(String),
    })
  })
})
