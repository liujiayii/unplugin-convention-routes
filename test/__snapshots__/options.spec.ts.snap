// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`options resolve > react 1`] = `
{
  "caseSensitive": false,
  "dirs": [
    {
      "baseRoute": "",
      "dir": "examples/react/src/pages",
    },
  ],
  "exclude": [
    "node_modules",
    ".git",
    "**/__*__/**",
  ],
  "extendRoute": undefined,
  "extensions": [
    "tsx",
    "jsx",
    "ts",
    "js",
  ],
  "extensionsRE": /\\\\\\.\\(tsx\\|jsx\\|ts\\|js\\)\\$/,
  "importMode": [Function],
  "importPath": "relative",
  "moduleIds": [
    "~react-pages",
    "virtual:generated-pages-react",
  ],
  "onClientGenerated": undefined,
  "onRoutesGenerated": undefined,
  "resolver": {
    "getComputedRoutes": [Function],
    "resolveExtensions": [Function],
    "resolveModuleIds": [Function],
    "resolveRoutes": [Function],
    "stringify": {
      "component": [Function],
      "dynamicImport": [Function],
      "final": [Function],
    },
  },
  "root": Any<String>,
  "routeBlockLang": "json5",
  "routeNameSeparator": "-",
  "routeStyle": "next",
}
`;

exports[`options resolve > solid 1`] = `
{
  "caseSensitive": false,
  "dirs": [
    {
      "baseRoute": "",
      "dir": "examples/solid/src/pages",
    },
  ],
  "exclude": [
    "node_modules",
    ".git",
    "**/__*__/**",
  ],
  "extendRoute": undefined,
  "extensions": [
    "tsx",
    "jsx",
    "ts",
    "js",
  ],
  "extensionsRE": /\\\\\\.\\(tsx\\|jsx\\|ts\\|js\\)\\$/,
  "importMode": [Function],
  "importPath": "relative",
  "moduleIds": [
    "~solid-pages",
  ],
  "onClientGenerated": undefined,
  "onRoutesGenerated": undefined,
  "resolver": {
    "getComputedRoutes": [Function],
    "resolveExtensions": [Function],
    "resolveModuleIds": [Function],
    "resolveRoutes": [Function],
    "stringify": {
      "dynamicImport": [Function],
      "final": [Function],
    },
  },
  "root": Any<String>,
  "routeBlockLang": "json5",
  "routeNameSeparator": "-",
  "routeStyle": "next",
}
`;

exports[`options resolve > vue - custom module id 1`] = `
{
  "caseSensitive": false,
  "dirs": [
    {
      "baseRoute": "",
      "dir": "examples/vue/src/pages",
    },
  ],
  "exclude": [
    "node_modules",
    ".git",
    "**/__*__/**",
  ],
  "extendRoute": undefined,
  "extensions": [
    "vue",
    "ts",
    "js",
  ],
  "extensionsRE": /\\\\\\.\\(vue\\|ts\\|js\\)\\$/,
  "importMode": [Function],
  "importPath": "relative",
  "moduleIds": [
    "~vue-pages",
  ],
  "onClientGenerated": undefined,
  "onRoutesGenerated": undefined,
  "resolver": {
    "getComputedRoutes": [Function],
    "hmr": {
      "added": [Function],
      "changed": [Function],
      "removed": [Function],
    },
    "resolveExtensions": [Function],
    "resolveModuleIds": [Function],
    "resolveRoutes": [Function],
  },
  "root": Any<String>,
  "routeBlockLang": "json5",
  "routeNameSeparator": "-",
  "routeStyle": "next",
}
`;

exports[`options resolve > vue 1`] = `
{
  "caseSensitive": false,
  "dirs": [
    {
      "baseRoute": "",
      "dir": "examples/vue/src/pages",
    },
  ],
  "exclude": [
    "node_modules",
    ".git",
    "**/__*__/**",
  ],
  "extendRoute": undefined,
  "extensions": [
    "vue",
    "ts",
    "js",
  ],
  "extensionsRE": /\\\\\\.\\(vue\\|ts\\|js\\)\\$/,
  "importMode": [Function],
  "importPath": "relative",
  "moduleIds": [
    "~pages",
    "pages-generated",
    "virtual:generated-pages",
  ],
  "onClientGenerated": undefined,
  "onRoutesGenerated": undefined,
  "resolver": {
    "getComputedRoutes": [Function],
    "hmr": {
      "added": [Function],
      "changed": [Function],
      "removed": [Function],
    },
    "resolveExtensions": [Function],
    "resolveModuleIds": [Function],
    "resolveRoutes": [Function],
  },
  "root": Any<String>,
  "routeBlockLang": "json5",
  "routeNameSeparator": "-",
  "routeStyle": "next",
}
`;
